### ISSUE TITLE
Limit checks are missing

### ISSUE THREAD

**Issue by wilson-ko** [stulp/dmpbbo#23]
I found that this library does not consider constraints such as acceleration, velocity or position limits. I did find the forcing term scaling here:

https://github.com/stulp/dmpbbo/blob/master/src/dmp/Dmp.cpp#L629

But it doesn't seem to prevent the system from breaching its limits. I'm currently unsure whether the feature is not implemented yet or whether I just didn't find it.

I think that such a (optional) feature can be easily implemented by introducing a `limits_` member in `DynamicalSystem` and performing an element-wise `std::max` in `integrateStep`. Do you think that's something we should implement? If so I will create a PR for this.

**Comment by stulp** [comment]
Thanks for these considerations. From my point of view, the (dynamical) system itself has no limits to breach. The robot system may have such limits, but then I would rather implement this on the robot side and not on the DMP side. Why? Say I have an RRT motion planner, a DMP, and a spline-based motion generator. It's easier to implement the limits once on the robot so that they are always enforced, rather than writing a separate function in each of the motion generators. Does that make sense?

If you have good arguments for making it part of the Dmp and not the robot, please let me know (after all this feature wouldn't hurt users that do not need it).

**Comment by wilson-ko** [comment]
First of all, thanks for the quick response. I see, that does make sense and I totally agree. However I think that my most suitable option is to still implement the limits on the DMP side.. I could be wrong but I'm curious about your opinion, so let me try to summarize my problem.

I am working with a robot arm that accepts position waypoints as input in real-time (every 8ms). So basically I need to provide the next position within 8ms in an online fashion, for which I want to use a DMP as the only motion generator (so the DMP directly provides the next position waypoint to the controller by taking an integration step).

If the next position is too far from its current position, the robot needs to accelerate more, causing the robot to throw an error and stop moving (this is also true if the jerk, velocity or position bounds are violated). Since the DMP directly provides the waypoints to the controller, it needs to provide the position waypoints such that all these limits are respected. The problem is also that I don't have access to the internals of the robot controller, so I believe that in my case the DMP is the lowest level of access I have.

Of course I could implement something in between the DMP and the controller that keeps track of the current position and calculates the derivatives which I can cap before sending them to the controller. Then whenever I detect a breach I'd need to bring the next waypoint closer to the current one such that the limits are not violated. However, I believe that it's more natural to implement this in the DMP, because I'd immediately get the correct waypoint if I cap the accelerations and velocities before I take the integration step. And since it's an optional feature other users who don't need it wouldn't notice anything, but it could be useful for other people with a similar problem.

If you think it's still better for me to implement this outside the DMP then I'll give that a try, but I'm curious about your opinion on this since you may have other ideas on this matter. Thanks and I'm looking forward to your response!

**Comment by stulp** [comment]
OK, I understand. Let me elaborate, it's a bit tricky, and it's probably not the answer you're hoping for.

* A DMP is a second order dynamical system, which can be rewritten as first order dynamical system \dot{x} = f(x), see http://freekstulp.net/dmpbbo/html/page_dyn_sys.html#dyn_sys_rewrite_second_first and http://freekstulp.net/dmpbbo/html/page_dmp.html. This rewrite means that \dot{x} contains velocities AND accelerations of the same variables. If you put limits on some of the variables, but do not update the others accordingly, you may violate the relationship defined by the differential equation of the dynamical system, and inconsistencies between the variables will arise. This may cause bad discontinuities, and the stability of the system can no longer be guaranteed. That may cause far worse problems than the ones your are trying to solve!

* As DMPs are dynamical systems, their output is inherently in terms of changes in variables (velocities, accelerations), not in terms of the variables themselves (positions). You can get the positions by integrating them out, but they are not the "raw output". So limiting the positions in the DMP is not  possible (?) A dynamical way of thinking about limits is to add a forcing term that represents a repellor (in acceleration space) away from these limits. This ensures the movement remains smooth, and that the variables are consistent.

* For the reason above, a position-controlled robot is not really ideal to exploit the nice features of a DMP, because DMPs define the movement in acceleration space, not position space. And controlling in acceleration space usually means you need to have torque-control access to the robot. On several occasions, I've seen students getting ugly jerky movements with DMPs on position-controlled robots, because there is some controller in between doing some nasty interpolation. I fear you will get the same, especially when limits are enforced. For a position-controlled robot, you might as well use a (static) spline representation, in which such limits are much easier to enforce. 

On a poetic note: The beauty and expressiveness of dance is due to its fluent, dynamic nature. Watching static pictures of dance far less satisfying. The same goes for DMPs: let them express themselves through accelerations! And not by integrating out static positions and passing these to some interpolator that converts back to accelerations internally in some unknown way.

Finally, I would be OKish with placing limits on the acceleration within the DMP. This is basically like adding another forcing term. But even this I would be wary of, because the forcing term is usually smooth, whereas a hard cap can introduce discontinuities. Let me try and implement this.

**Comment by stulp** [comment]
I implemented a first quick version in a branch, see b431ba7facf999

Of course, if you limit the accelerations, the path in position space is no longer the same (try calling testDmpAccelerationLimitsWrapper.py after make install), because the DMP does not have an explicit representation of the trajecory in position space. 

So I don't think this should be merged into master, because it causes more trouble than it's worth, in my opinion.

**Comment by wilson-ko** [comment]
I took a look at it and seems like I implemented something similar, except then in the DynamicalSystem class. Yeah actually my conclusion was also the same that it's more trouble than it's worth indeed, it seems like the limits are experienced as some kind of perturbation against which the system tries to compensate against.

Anyway thanks for considering this and making a quick test, since we both conclude that it's not worth the trouble we should close this ticket. I will try to solve my problem in a different way.
